post: resolving Git conflicts with a merge driver 🛂 (#53)
cloudx-labs/posts: posts/navarroaxel/09.resolving-git-conflicts-using-a-merge-driver.md

---
title: Resolving Git conflicts in package-lock.json using a merge driver
published: true
description: Using Git attributes and merge drivers to resolve conflicts automatically.
tags: 'git, npm, node, github'
cover_image: ./assets/resolving-git-conflicts-merge-driver-cover.jpg
id: 1188285
---

> Everyone that uses auto generated files like the `package-lock.json`, `yarn.lock`, `Gemfile.lock` or any stores code coverage reports (or test results) in XML format inside a Git repository has experienced Git conflicts in these files.

Reviewer: ```suggestion
Everyone that uses auto generated files like the `package-lock.json`, `yarn.lock`, `Gemfile.lock` or any files that store code coverage reports (or test results) in XML format inside a Git repository has experienced Git conflicts in these files.
```

> Let's check how to resolve this kind of conflicts in a smoother way.

Reviewer: ```suggestion
Let's see how to resolve this kind of conflicts in a smoother way.
```

I'm sure you know that you can delete the `package-lock.json` file and run `npm install` again - or similar with Yarn.

```bash
rm package-lock.json
npm install
git add package-lock.json
```

But Git can resolve this automatically (plus this solution works for other auto generated files and binary files, e.g. images).

## The Git attributes

> You can create a special file called [`.gitattributes`](https://git-scm.com/docs/gitattributes) placed in the root of your Git repository where you can manually configure how Git handles files. Let's check an example:

Reviewer: ```suggestion
You can create a special file called [`.gitattributes`](https://git-scm.com/docs/gitattributes) placed in the root of your Git repository where you can manually configure how Git handles files. Let's see an example:
```

```bash
git checkout –theirs package-lock.json
```

This can also be done with a custom merge driver. 👐

## A Git merge driver

A merge driver tries to resolve Git conflicts and you can [create your own](https://git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver) one. Let's define a driver called `theirs` that will use the upstream version (`%B`) over our version (`%A`).

```bash
git config --global merge.theirs.name "Keep upstream changes"
git config --global merge.theirs.driver "cp -f '%B' '%A'"
```

💡 Use `--global` to make this driver available to all repositories in your machine.

Then, add the following to your `.gitattributes` (or `.git/info/attributes`, or `~/.config/git/attributes`) file.

```text
package-lock.json merge=theirs
```

Now, every conflict will be resolved by our driver accepting the upstream version. Remember that you should run `npm install` to re-add your changes to the `package-lock.json` because this driver is not smart enough, and discards your changes to the lock file. You need to keep the file in sync with your changes to the `package.json`.

### The npm merge driver

> The npm team has created a [npm-merge-drive](https://github.com/npm/npm-merge-driver) which makes a smarter merge than our previous driver - which does an overwrite 😁, keeping the lock file in sync with our changes in the `package.json`. You can just install it globally in your machine without affecting your coworkers because the `~/.config/git/attributes` is used for all your cloned repositories.

Reviewer: ```suggestion
The npm team has created an [npm-merge-drive](https://github.com/npm/npm-merge-driver) which makes a smarter merge than our previous driver - which does an overwrite 😁, keeping the lock file in sync with our changes in the `package.json`. You can just install it globally in your machine without affecting your coworkers because the `~/.config/git/attributes` is used for all your cloned repositories.
```

```bash
npx npm-merge-driver install --global
```

The bad part is that this repo is archived and doesn't receive any updates. 🐛

> You can find other merge drivers in the npm registry, even for another type of files like the [git-json-merge](https://github.com/jonatanpedersen/git-json-merge).

Reviewer: ```suggestion
You can find other merge drivers in the npm registry, even for other types of files like the [git-json-merge](https://github.com/jonatanpedersen/git-json-merge).
```
